### PR TITLE
Calculated Summary Page

### PR DESCRIPTION
**Pull request by SamGodwin2**
### What is the context of this PR?

This PR contains all the work required to implement the new calculated summary page on the front-end api and publisher. 

### How to review 
Tests pass and the new feature looks and works as expected.

### CODE REVIEW

**Comment by SamGodwin2** [comment]
[Calculated summary (12)](https://trello.com/c/IjLzPdVy/704-calculated-summary-12)

**Comment by hamishtaplin** [comment]
Duplicating a section with a calc summary crashes:

Screenshot 2019-04-11 14 33 29

**Comment by hamishtaplin** [comment]
Page title piping content picker should default to "variables".

File: eq-author-api/schema/resolvers/base.js [hunk 1]
```diff
@@ -27,9 +23,22 @@ const {
...
 const { currentVersion } = require("../../migrations");
 
+const {
+  getSection,
+  getPage,
+  getAnswers,
+  getAnswer,
+  getOption,
+  getValidation,
+  getAvailableMetadataForValidation,
```
Comment (tgandrews) [inline eq-author-api/schema/resolvers/base.js]: A few of these don't sound re-used. Like `getAvailableMetadataForValidation`, `getAvailablePreviousAnswersForValidation` and `getQuestionnaireList`.

Can we only put reused things in the utils.

File: eq-author-api/schema/resolvers/pages/calculatedSummaryPage.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    ctx
+  ) => {
+    const section = find(ctx.questionnaire.sections, {
+      id: sectionId,
+    });
+    const page = createCalculatedSummary({ sectionId });
+    const insertionPosition =
+      typeof position === "number" ? position : section.pages.length;
+    section.pages.splice(insertionPosition, 0, page);
+    page.position = insertionPosition;
```
Comment (tgandrews) [inline eq-author-api/schema/resolvers/pages/calculatedSummaryPage.js]: This shouldn't saved it is a calculated property.

File: eq-author-api/schema/tests/calculatedSummary.test.js [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+                },
+                {
+                  type: CURRENCY,
+                },
+              ],
+            },
+            {
+              pageType: "calculatedSummary",
+            },
+            {
```
Comment (tgandrews) [inline eq-author-api/schema/tests/calculatedSummary.test.js]: Is this page ever used?

File: eq-author-api/schema/tests/calculatedSummary.test.js [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    expect(result).toMatchObject({
+      id: expect.any(String),
+      availableSummaryAnswers: [
+        { id: answersPage.answers[1].id },
+        { id: answersPage.answers[2].id },
+      ],
+    });
+  });
+
+  it("should return more than one type of summary answeres if more than one type selected", async () => {
```
Comment (tgandrews) [inline eq-author-api/schema/tests/calculatedSummary.test.js]:
```suggestion
  it("should return more than one type of summary answers if more than one type selected", async () => {
```

File: eq-author-api/tests/utils/questionnaireBuilder/index.js [hunk 5]
```diff
@@ -48,56 +51,62 @@ const buildQuestionnaire = async questionnaireConfig => {
...
-            sectionId: createdSection.id,
-            ...page,
-          });
-          if (page.confirmation) {
-            const createdQuestionConfirmation = await createQuestionConfirmation(
-              questionnaire,
-              {
-                pageId: createdPage.id,
-                ...page.confirmation,
+          if (get(page, "pageType") === "calculatedSummary") {
```
Comment (tgandrews) [inline eq-author-api/tests/utils/questionnaireBuilder/index.js]: Why are we using `get` here?

File: eq-author/src/App/QuestionnaireDesignPage/index.test.js [hunk 6]
```diff
@@ -160,6 +161,78 @@ describe("QuestionnaireDesignPage", () => {
...
+        },
+      });
+
+      wrapper.find(NavigationSidebar).simulate("addQuestionPage");
+
+      wrapper.find(NavigationSidebar).simulate("addCalculatedSummaryPage");
+
+      expect(mockHandlers.onAddCalculatedSummaryPage).toHaveBeenCalledWith(
+        section.id,
+        1
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnaireDesignPage/index.test.js]: This isn't the start of the section as described in the test title so I'm not sure what this is proving

File: eq-author/src/App/QuestionnaireDesignPage/index.test.js [hunk 7]
```diff
@@ -160,6 +161,78 @@ describe("QuestionnaireDesignPage", () => {
...
+
+      wrapper.find(NavigationSidebar).simulate("addCalculatedSummaryPage");
+
+      expect(mockHandlers.onAddCalculatedSummaryPage).toHaveBeenCalledWith(
+        section.id,
+        0
+      );
+    });
+
+    it("should be able to add a page after the confirmation when on a confirmation page", () => {
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnaireDesignPage/index.test.js]: Confirmation doesn't ever have any effect on this in author. This test IMO is not worthwhile.

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+  margin-bottom: 1em;
+`;
+
+const NoShortcuts = styled.div`
+  padding: 0.5em 1em;
+  font-weight: normal;
+  font-size: 0.9em;
+`;
+
+const translateSelectedAnswers = selectedAnswers => {
```
Comment (tgandrews) [inline eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js]: As you mentioned. These answers can no longer go across multiple sections so this isn't needed anymore.

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+const PageSegment = styled.div`
+  padding: 0 2em;
+`;
+
+const SelectorTitle = styled.h2`
+  font-size: 1em;
+  color: ${colors.darkGrey};
+`;
+
+export class CalculatedSummaryPageEditor extends React.Component {
```
Comment (tgandrews) [inline eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js]: This could be a functional component as it implements render. This could possibly be a lint.
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  onChangeUpdate: PropTypes.func.isRequired,
+  onUpdateCalculatedSummaryPage: PropTypes.func.isRequired,
+};
+
+export default flowRight(
+  withUpdateCalculatedSummaryPage,
+  withPropRenamed("onUpdateCalculatedSummaryPage", "onUpdate"),
+  withEntityEditor(
+    "page",
+    CalculatedSummaryPageEditor.fragments.CalculatedSummaryPage
```
Comment (tgandrews) [inline eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js]: withEntityEditor does not need a fragment.

File: eq-author/src/enhancers/withCreateCalculatedSummaryPage.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import createCalculatedSummaryPageMutation from "graphql/createCalculatedSummaryPage.graphql";
+import { buildPagePath } from "utils/UrlUtils";
+import { get, tap } from "lodash/fp";
+
+export const redirectToNewPage = ({ history, match: { params } }) => page => {
+  const { id, section } = page;
+  history.push(
+    buildPagePath({
+      questionnaireId: params.questionnaireId,
+      sectionId: section.id,
```
Comment (tgandrews) [inline eq-author/src/enhancers/withCreateCalculatedSummaryPage.js]: The sectionId isn't needed.

File: eq-author-api/schema/resolvers/pages/index.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    return removedPage;
+  },
+  deletePage: async (_, { input }, ctx) => {
+    const section = findSectionByPageId(ctx.questionnaire.sections, input.id);
+    remove(section.pages, { id: input.id });
+    await saveQuestionnaire(ctx.questionnaire);
+    return section;
+  },
+
+  duplicatePage: async (_, { input }, ctx) => {
```
[the remark this answers is not in the document]
Comment (SamGodwin2) [reply eq-author-api/schema/resolvers/pages/index.js]: It's designed to (and can) be used on all types of pages it was just decided that duplicating a calsum page creates a number of weird edge cases. However the duplication mutation still feels like something that belongs to a page not just a questionPage.

File: eq-author-api/schema/tests/calculatedSummary.test.js (hunk 3 above)
Comment (SamGodwin2) [reply eq-author-api/schema/tests/calculatedSummary.test.js]: That last page is there to ensure that pages and their answers after the calSum aren't present in the availableAnswers. I've updated the test to make it more explicit and explain the test a bit more.

File: eq-author/src/App/QuestionnaireDesignPage/index.test.js (hunk 6 above)
Comment (SamGodwin2) [reply eq-author/src/App/QuestionnaireDesignPage/index.test.js]: Good question 🤷‍♂️

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+  piping: true,
+};
+
+const PageSegment = styled.div`
+  padding: 0 2em;
+`;
+
+const SelectorTitle = styled.h2`
+  font-size: 1em;
+  color: ${colors.darkGrey};
```
Comment (hamishtaplin) [inline eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js]:
```suggestion
  color: ${colors.darkGrey};
  margin: 0 0 0.4em;
```
